File: src/my_project/interviews/top_150_questions_round_22/ex_108_maximum_circular_subarray.py
from typing import List, Union, Collection, Mapping, Optional
from abc import ABC, abstractmethod
from collections import deque, defaultdict

class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        total = 0
        max_sum = nums[0]
        cur_max = 0
        min_sum = nums[0]
        cur_min = 0

        for num in nums:
            cur_max = max(cur_max + num, num)
            max_sum = max(max_sum, cur_max)
            cur_min = min(cur_min + num, num)
            min_sum = min(min_sum, cur_min)
            total += num

        if max_sum > 0:
            return max(max_sum, total - min_sum)
        return max_sum

File: src/my_project/interviews_typescript/top_150_questions_round_1/ex_108_maximum_circular_subarray.ts
function maxSubarraySumCircular(nums: number[]): number {
    let total = 0;
    let maxSum = nums[0];
    let curMax = 0;
    let minSum = nums[0];
    let curMin = 0;

    for (const num of nums) {
        curMax = Math.max(curMax + num, num);
        maxSum = Math.max(maxSum, curMax);
        curMin = Math.min(curMin + num, num);
        minSum = Math.min(minSum, curMin);
        total += num;
    }

    if (maxSum > 0) {
        return Math.max(maxSum, total - minSum);
    }
    return maxSum;
};